ucan/src/capability/data.rs

```diff
@@ -0,0 +1,190 @@
+/// form. Use `iter()` to deconstruct this map into a sequence of [Capability] datas.
+///
+/// ```
+/// use ucan::capability::{Capabilities, Capability};
+/// use serde_json::json;
+///
+/// let capabilities = Capabilities::try_from(&json!({
+///   "example://example.com/private/84MZ7aqwKn7sNiMGsSbaxsEa6EPnQLoKYbXByxNBrCEr": {
+///     "wnfs/append": [{}]
+///   },
+///   "mailto:username@example.com": {
+///     "msg/receive": [{ "max_count": 5 }],
+///     "msg/send": [{}]
+///   }
+/// })).unwrap();
+///
+/// assert_eq!(capabilities.iter().collect::<Vec<Capability>>(), vec![
+///   Capability::from(("example://example.com/private/84MZ7aqwKn7sNiMGsSbaxsEa6EPnQLoKYbXByxNBrCEr", "wnfs/append", &vec![json!({})])),
+///   Capability::from(("mailto:username@example.com", "msg/receive", &vec![json!({ "max_count": 5 })])),
+///   Capability::from(("mailto:username@example.com", "msg/send", &vec![json!({})])),
+/// ]);
+/// ```
+pub struct Capabilities(CapabilitiesImpl);
+
+impl Capabilities {
+    pub fn iter(&self) -> CapabilitiesIterator {
+        self.0
+            .iter()
+            .flat_map(|(resource, abilities): (&String, &AbilitiesImpl)| {
+                abilities
+                    .iter()
+                    .map(Box::new(|(ability, caveats): (&String, &Vec<Value>)| {
+                        Capability::new(resource.to_owned(), ability.to_owned(), caveats.to_owned())
+                    }))
+            })
+    }
+}
+
+impl Deref for Capabilities {
+    type Target = CapabilitiesImpl;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl TryFrom<Vec<Capability>> for Capabilities {
+    type Error = anyhow::Error;
+    fn try_from(value: Vec<Capability>) -> Result<Self, Self::Error> {
+        let mut resources: CapabilitiesImpl = BTreeMap::new();
+        for capability in value.into_iter() {
+            let (resource_name, ability, caveats) =
+                <(String, String, Vec<Value>)>::from(capability);
+
+            let resource = if let Some(resource) = resources.get_mut(&resource_name) {
+                resource
+            } else {
+                let resource: AbilitiesImpl = BTreeMap::new();
+                resources.insert(resource_name.clone(), resource);
+                resources.get_mut(&resource_name).unwrap()
+            };
+
+            for value in caveats.iter() {
+                if !value.is_object() {
+                    return Err(anyhow!("Caveat must be an object: {}", value));
+                }
+            }
+            resource.insert(ability, caveats);
+        }
+        Capabilities::try_from(resources)
+    }
+}
+
+impl TryFrom<CapabilitiesImpl> for Capabilities {
+    type Error = anyhow::Error;
+
+    fn try_from(value: CapabilitiesImpl) -> Result<Self, Self::Error> {
+        for (resource, abilities) in value.iter() {
+            if abilities.is_empty() {
+                // 0.10.0: 3.2.6.2 One or more abilities MUST be given for each resource.
```

Comment:
A permalink to the relevant section of the spec on Github would be good to include in places like this